### PR TITLE
RavenDB-16510 Add TCP connections to the traffic watch

### PR DESCRIPTION
**Pull request by shiranshalom**

### CODE REVIEW

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs [hunk 1]
```diff
@@ -352,19 +353,76 @@ public class IndexRenameChange : IndexChange
...
+            json[nameof(Source)] = Source;
+            json[nameof(Certificate)] = Certificate;
+            json[nameof(Operation)] = Operation;
+            json[nameof(OperationVersion)] = OperationVersion;
+        }
+    }
+
+    public abstract class TrafficWatchChangeBase : DatabaseChange, IDynamicJson
+    {
+        public virtual TrafficWatchType TrafficWatchType { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Changes/ChangeNotification.cs]: abstract?
only get needed?

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs [hunk 2]
```diff
@@ -352,19 +353,76 @@ public class IndexRenameChange : IndexChange
...
+        public DateTime TimeStamp { get; set; }
         public string DatabaseName { get; set; }
         public string CustomInfo { get; set; }
-        public TrafficWatchChangeType Type { get; set; }
         public string ClientIP { get; set; }
+
+        public DynamicJsonValue ToJson()
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Changes/ChangeNotification.cs]: virtual?

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs [hunk 3]
```diff
@@ -352,19 +353,76 @@ public class IndexRenameChange : IndexChange
...
+                [nameof(DatabaseName)] = DatabaseName,
+                [nameof(CustomInfo)] = CustomInfo,
+                [nameof(ClientIP)] = ClientIP
+            };
+
+            ToJson(json);
+            return json;
+        }
+
+        protected abstract void ToJson(DynamicJsonValue json);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Changes/ChangeNotification.cs]: not needed if main method marked as virtual and overridden

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs [hunk 4]
```diff
@@ -352,19 +353,76 @@ public class IndexRenameChange : IndexChange
...
+    }
+
+    public enum TrafficWatchType
+    {
+        TrafficWatchHttpChange,
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Changes/ChangeNotification.cs]: rename to `Http` and `Tcp` ?

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs [hunk 5]
```diff
@@ -352,19 +353,76 @@ public class IndexRenameChange : IndexChange
         public string OldIndexName { get; set; }
     }
 
-    public class TrafficWatchChange : DatabaseChange
+    public class TrafficWatchHttpChange : TrafficWatchChangeBase
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Changes/ChangeNotification.cs]: mark everything as internal

File: src/Raven.Server/RavenServer.cs [hunk 6]
```diff
@@ -1882,16 +1889,22 @@ private void ListenToNewTcpConnection(TcpListener listener)
...
-                            if (await DispatchServerWideTcpConnection(tcp, header, buffer))
+                            if (await DispatchServerWideTcpConnection(tcp, header, buffer))  
                             {
                                 tcp = null; //do not keep reference -> tcp will be disposed by server-wide connection handlers
                                 return;
                             }
 
                             await DispatchDatabaseTcpConnection(tcp, header, buffer, cert);
+
+                            if (TrafficWatchManager.HasRegisteredClients)
```
Comment (ppekrol) [inline src/Raven.Server/RavenServer.cs:1921]: What will happen if we have a server-wide, test or ping connection?

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Changes/ChangeNotification.cs]: Done.

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs (hunk 4 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Changes/ChangeNotification.cs]: Done.

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs (hunk 1 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Changes/ChangeNotification.cs]: Done.

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs [hunk 7]
```diff
@@ -352,19 +353,76 @@ public class IndexRenameChange : IndexChange
         public string OldIndexName { get; set; }
     }
 
-    public class TrafficWatchChange : DatabaseChange
+    internal class TrafficWatchHttpChange : TrafficWatchChangeBase
```
Comment (ayende) [inline src/Raven.Client/Documents/Changes/ChangeNotification.cs:291]: This is a breaking change, no?

File: src/Raven.Server/RavenServer.cs [hunk 8]
```diff
@@ -1931,6 +1933,25 @@ private void ListenToNewTcpConnection(TcpListener listener)
             });
         }
 
+        private async Task DispatchTcpConnection(TcpConnectionHeaderMessage header, TcpConnectionOptions tcp, JsonOperationContext.MemoryBuffer buffer, X509Certificate2 cert)
+        {
+            if (header.Operation == TcpConnectionHeaderMessage.OperationTypes.TestConnection ||
+                header.Operation == TcpConnectionHeaderMessage.OperationTypes.Ping)
+            {
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs:1940]: Need to notify here, no?

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs (hunk 7 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Changes/ChangeNotification.cs:291]: Yes, but this is not used anywhere in the Client API

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Changes/ChangeNotification.cs]: Done.

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs (hunk 3 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Changes/ChangeNotification.cs]: Done.

**Review by ayende — APPROVED** [review]

**Review by ppekrol — APPROVED** [review]

**Comment by ppekrol** [comment]
Requires Studio work cc @ml054

**Comment by ppekrol** [comment]
@shiranshalom Can you move it to v4.2?

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs [hunk 9]
```diff
@@ -256,19 +257,77 @@ public class IndexRenameChange : IndexChange
...
-
-    public class TrafficWatchChange : DatabaseChange
+    internal abstract class TrafficWatchChangeBase : DatabaseChange, IDynamicJson
     {
+        public abstract TrafficWatchType TrafficWatchType { get; }
         public DateTime TimeStamp { get; set; }
+        public string DatabaseName { get; set; }
+        public string CustomInfo { get; set; }
+        public string ClientIP { get; set; }
+        public string Thumbprint { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Changes/ChangeNotification.cs]: CertificateThumbprint - it needs to be clear what thumbprint it is

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Client/Documents/Changes/ChangeNotification.cs (hunk 9 above)
Comment (ml054) [reply src/Raven.Client/Documents/Changes/ChangeNotification.cs]: @shiranshalom after you fix server side, please assign issue to me - I will refactor studio part.

Comment (shiranshalom) [reply src/Raven.Client/Documents/Changes/ChangeNotification.cs]: Done.

**Comment by ml054** [comment]
fixed studio compilation

**Pull request merged**